pulsar-broker/src/main/java/org/apache/pulsar/broker/service/persistent/PersistentSubscription.java

```diff
@@ -209,6 +276,87 @@ public void acknowledgeMessage(List<Position> positions, AckType ackType, Map<St
+     * smaller than that one it cumulative acked, it'll succeed.
+     * <p>
+     * If transaction is aborted all messages acked by it will be put back to pending state.
+     *
+     * @param txnID                  TransactionID of an ongoing transaction trying to sck message.
+     * @param positions              {@link Position}(s) it try to ack.
+     * @param ackType                {@link AckType}.
+     * @throws TransactionConflictException
+     * @throws IllegalArgumentException
+     */
+    public synchronized void acknowledgeMessage(TxnID txnID, List<Position> positions, AckType ackType) throws TransactionConflictException {
+        checkArgument(txnID != null, "TransactionID can not be null.");
+        if (AckType.Cumulative == ackType) {
+            // Check if another transaction is already using cumulative ack on this subscription.
+            if (this.txnID != null && this.txnID != txnID) {
+                String errorMsg = "[" + topicName + "][" + subName + "] Transaction:" + txnID.getMostSigBits() +
+                        txnID.getLeastSigBits() + " try to cumulative ack message while transaction:"
+                        + this.txnID.getMostSigBits() + this.txnID.getLeastSigBits()
+                        + " already cumulative acked messages.";
+                log.error(errorMsg);
+                throw new TransactionConflictException(errorMsg);
+            }
+
+            if (positions.size() != 1) {
+                String errorMsg = "[" + topicName + "][" + subName + "] Transaction:" + txnID.getMostSigBits() +
+                txnID.getLeastSigBits() + " invalid cumulative ack received with multiple message ids.";
+                log.error(errorMsg);
+                throw new IllegalArgumentException(errorMsg);
+            }
+
+            Position position = positions.get(0);
+
+            if (log.isDebugEnabled()) {
+                log.debug("[{}][{}] TxnID:[{}{}] Cumulative ack on {}.", topicName, subName, txnID.getMostSigBits(),
+                        txnID.getLeastSigBits(), position);
+            }
+
+             if (this.txnID == null) {
+                // Only set txnID if no transaction is doing cumulative ack.
+                TXNID_UPDATER.set(this, txnID);
+                POSITION_UPDATER.set(this, position);
+            } else if (((PositionImpl)position).compareTo((PositionImpl)this.pendingCumulativeAckMessage) > 0) {
+                // If new cumulative ack position is greater than current one, update it.
+                POSITION_UPDATER.set(this, position);
+            }
+        } else {
+            if (log.isDebugEnabled()) {
+                log.debug("[{}][{}] TxnID:[{}{}] Individual acks on {}", topicName, subName, txnID.getMostSigBits(),
+                        txnID.getLeastSigBits(), positions);
+            }
+
+            ConcurrentOpenHashSet<Position> pendingAckMessage =
```

Comment:
I would suggest renaming this variable to `pendingAckMessagesCurTxn`. Otherwise the name is too close to `pendingAckMessages`.

Reply:
renamed to pendingAckMessageForCurrentTxn.